<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="Data representation and clustering of diseases and traits : Code for HyperBrowser tools of clustering of diseases and traits, as represented data from Genome-Wide Association Studies (GWAS)">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>Data representation and clustering of diseases and traits</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/johhorn/gwas-clustering">View on GitHub</a>

          <h1 id="project_title">Tools for representation and clustering of diseases and traits</h1>
          <h2 id="project_tagline">Implementation of the master thesis by Johanne H. Horn, carried out at the Biomedical Informatics (BMI) research group at the Department of Informatics, University of Oslo.</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/johhorn/gwas-clustering/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/johhorn/gwas-clustering/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h3>
<a id="code-for-clustering-of-diseases-and-traits" class="anchor" href="#code-for-clustering-of-diseases-and-traits" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Code for clustering of diseases and traits</h3>
<p>
  This page describes the GitHub repository for the master thesis "Tools for cluster analysis of data from genome-wide association studies: Role of data representation and similarity measures", written by Johanne H. Horn at the research group of Biomedical Informatics (BMI), University of Oslo (UiO).
</p>
<p>
  The majority of the code available in this repository is part of the Genomic HyperBrowser, an open source, web-based software system for comparative genomic analysis. The HyperBrowser is developed and supported by the BMI research group at UiO and the Radium Hospital in Oslo. The online interface can be found <a href="http://hyperbrowser.uio.no/clustering/">here</a>, where the following tools are provided:</p>

  <h6>Clustering of binary data representations </h6>
  <ul>
    <li>Binary vector clustering</li>
    <li>Genetic loci clustering</li>
    <li>Haplotype block clustering</li>
    <li>Bipartite matching of point tracks</li>
  </ul>
  <h6>Clustering of continuous vectors</h6>
  <ul>
    <li>Correlation of track values</li>
    <li>Correlation of overlap with other tracks</li>
  </ul>
  <h6>Empirical exploration tools</h6>
  <ul>
    <li>Empirical properties of point tracks</li>
    <li>Empirical properties of LD tracks</li>
    <li>Print rsids of tracks in GSuite</li>
  </ul>
  <h6>Create and modify data</h6>
  <ul>
    <li>GSuite liftover</li>
    <li>Convert GSuite sumstat tracks</li>
    <li>LD GSuite generator</li>
    <li>LD track generator</li>
  </ul>

<p>
  Most of the code provided cannot be run as standalone scripts, as they rely on the HyperBrowser framework, in particular the files of <a href="https://github.com/johhorn/gwas-clustering/tree/master/Tools">tools</a>, <a href="https://github.com/johhorn/gwas-clustering/tree/master/Statistics">statistics</a> and <a href="https://github.com/johhorn/gwas-clustering/tree/master/Helper%20classes">general helper classes</a>.
  These tools and statistics provide different implementations for feature representation and comparison of genomic tracks. The tracks represent diseases and traits, and represented by associated variants from Genome-Wide Association Studies (GWAS).
  In addition to HyperBrowser code, scripts for <a href="https://github.com/johhorn/gwas-clustering/tree/master/LD%20scripts">LD computation</a> for variants, as given by their rsids, is provided.
</p>

      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">Data representation and clustering of diseases and traits maintained by <a href="https://github.com/johhorn">johhorn</a></p>
        <p>Published with <a href="https://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

    

  </body>
</html>
